File: day03/main.go
package main

import (
	"fmt"
	"math"
	"os"
	"strings"
)

func countZeroOnes(vals []string) ([]int, []int) {
	size := len(vals[0])
	ones := make([]int, size)
	zeros := make([]int, size)
	for _, val := range vals {
		for i, c := range val {
			if c == '1' {
				ones[i] += 1
			} else {
				zeros[i] += 1
			}
		}
	}
	return ones, zeros
}

func filterValues(vals []string, c string, pos int) []string {
	res := make([]string, 0)
	for _, v := range vals {
		if string(v[pos]) == c {
			res = append(res, v)
		}
	}
	return res
}

func filterEmpty(s []string) []string {
	var r []string
	for _, s := range s {
		if strings.TrimSpace(s) != "" {
			r = append(r, s)
		}
	}
	return r
}

func toInt(s string) int {
	r := 0
	for i, c := range s {
		if string(c) == "1" {
			r += int(math.Pow(2, float64(len(s)-i-1)))
		}
	}
	return r
}

func selectStr(vals []string, mostCommon bool) string {
	filtered := vals
	for i := 0; i < len(vals[0]); i++ {
		ones, zeros := countZeroOnes(filtered)
		check := ones[i] >= zeros[i]
		if !mostCommon {
			check = !check
		}
		c := "1"
		if !check {
			c = "0"
		}
		filtered = filterValues(filtered, c, i)
		if len(filtered) == 1 {
			break
		}
	}
	return filtered[0]
}

func main() {
	contents, _ := os.ReadFile(os.Args[1])
	vals := strings.Split(string(contents), "\n")
	vals = filterEmpty(vals)
	fmt.Println(toInt(selectStr(vals, true)) * toInt(selectStr(vals, false)))
}

File: day03/src/main.rs
fn filter_values<'a>(vals: &Vec<&'a str>, c: char, pos: usize) -> Vec<&'a str> {
    let mut r: Vec<&str> = Vec::new();
    for v in vals
        .into_iter()
        .filter(|s| s.chars().nth(pos).unwrap() == c)
    {
        r.push(v)
    }
    r
}

fn count_zero_ones(vals: &[&str]) -> (Vec<i32>, Vec<i32>) {
    let size = vals[0].len();

    let mut ones = vec![0; size];
    let mut zeros = vec![0; size];

    for val in vals {
        for (i, c) in val.chars().enumerate() {
            match c {
                '0' => zeros[i] += 1,
                '1' => ones[i] += 1,
                _ => (),
            }
        }
    }
    (ones, zeros)
}

fn filter_vec<'a>(vals: &Vec<&'a str>, most_common: bool) -> Vec<&'a str> {
    let mut filtered = vals.to_vec();
    for i in 0..vals[0].len() {
        if filtered.len() == 1 {
            break;
        }
        let (ones, zeros) = count_zero_ones(&filtered);
        let mut check = ones[i] >= zeros[i];
        check = if most_common { check } else { !check };
        let char = if check { '1' } else { '0' };
        filtered = filter_values(&filtered, char, i);
    }
    filtered
}

fn to_int(s: &str) -> i32 {
    let mut r = 0;
    for (i, c) in s.chars().enumerate() {
        if c == '1' {
            r += 2_i32.pow((s.len() - i - 1) as u32);
        }
    }
    r
}

fn main() {
    use std::env;
    use std::fs;
    let args: Vec<String> = env::args().collect();

    let contents = fs::read_to_string(&args[1]).unwrap();
    let vals: Vec<&str> = contents.split("\n").filter(|s| !s.is_empty()).collect();

    println!(
        "{:?}",
        to_int(filter_vec(&vals, true)[0]) * to_int(filter_vec(&vals, false)[0])
    )
}
